### PR TITLE
Add support for custom formats

### PR DESCRIPTION
**Pull request by mbj**
### Acceptance

* [ ] `stack test --fast`.

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb, @snusnu This is ready for review.

File: src/OpenApi/Schema.hs [hunk 1]
```diff
@@ -195,15 +195,20 @@ newtype Enum = Enum [JSON.Value]
...
+data Format
+  = CustomFormat Text  -- Non standard formats, explicitly allowed by OAS.
+  | UnixTime
+  deriving stock (Eq, Show)
 
 instance ToText Format where
   toText = \case
-    UnixTime -> "unix-time"
+    UnixTime            -> "unix-time"
+    CustomFormat format -> format
```
Comment (dkubb) [inline src/OpenApi/Schema.hs:205]: Should these lines be sorted?

Comment (mbj) [reply src/OpenApi/Schema.hs:205]: yes, but we wanted to avoid question style unless its an actual question.

Comment (dkubb) [reply src/OpenApi/Schema.hs:205]: There may be a semantic reason to leave it as-is, to correspond to the `FromJSON` instance, which is why I asked.

Comment (mbj) [reply src/OpenApi/Schema.hs:205]: ahh okay, no it can be sorted alphabetically as the pattern matches are unique individually. The order of patern matches only matters of one can be shadowed by the other. Which is not the case here was we are matching individual constructors that are mutually exclusive.

File: test/Main.hs [hunk 2]
```diff
@@ -14,9 +14,25 @@ import qualified Data.HashMap.Strict       as HashMap
...
+  where
+    accepted :: [TestTree]
+    accepted = mkAccepted <$>
+      [ ("decimal",   Schema.CustomFormat "decimal")
+      , ("unix-time", Schema.UnixTime)
+      ]
+
+    rejected :: [TestTree]
+    rejected = mkRejected (JSON.parseJSON @Schema.Format) "Error in $:" <$>
+      [(JSON.object [], "expected format, encountered Object")]
```
Comment (dkubb) [inline test/Main.hs]: The `[]` should be replaced with `empty` or `mempty`.

File: src/OpenApi/Schema.hs [hunk 3]
```diff
@@ -195,15 +195,20 @@ newtype Enum = Enum [JSON.Value]
   deriving newtype (JSON.FromJSON, JSON.ToJSON)
   deriving stock   (Eq, Show)
 
-data Format = UnixTime
-  deriving stock (Eq, GHC.Bounded, GHC.Enum, Show)
+data Format
+  = CustomFormat Text  -- Non standard formats, explicitly allowed by OAS.
+  | UnixTime
+  deriving stock (Eq, Show)
```
Comment (dkubb) [inline src/OpenApi/Schema.hs:201]: I noticed the `GHC.Bounded` and `GHC.Enum` instances were dropped too. Should this removal be extracted into a separate commit? I assume nothing in this project depends on it, otherwise the code would not compile.

Comment (mbj) [reply src/OpenApi/Schema.hs:201]: No, as the reason we remove these constraints lie in the addition of the `CustomFormat` constructor that is not enumeratable as it takes any `Text`.

**Review by dkubb — APPROVED** [review]
